packages/firestore/src/remote/watch_change.ts

```diff
@@ -409,16 +411,106 @@ export class WatchChangeAggregator {
         }
       } else {
         const currentSize = this.getCurrentDocumentCountForTarget(targetId);
+        // Existence filter mismatch. Mark the documents as being in limbo, and
+        // raise a snapshot with `isFromCache:true`.
         if (currentSize !== expectedCount) {
-          // Existence filter mismatch: We reset the mapping and raise a new
-          // snapshot with `isFromCache:true`.
-          this.resetTarget(targetId);
-          this.pendingTargetResets = this.pendingTargetResets.add(targetId);
+          // Apply bloom filter to identify and mark removed documents.
+          const bloomFilterApplied = this.applyBloomFilter(
+            watchChange.existenceFilter,
+            targetId,
+            currentSize
+          );
+          if (!bloomFilterApplied) {
+            // If bloom filter application fails, we reset the mapping and
+            // trigger re-run of the query.
+            this.resetTarget(targetId);
+            this.pendingTargetResets = this.pendingTargetResets.add(targetId);
+          }
         }
       }
     }
   }
 
+  /** Returns whether a bloom filter removed the deleted documents successfully. */
+  private applyBloomFilter(
+    existenceFilter: ExistenceFilter,
+    targetId: number,
+    currentCount: number
+  ): boolean {
+    const unchangedNames = existenceFilter.unchangedNames;
+    const expectedCount = existenceFilter.count;
+
+    if (!unchangedNames || !unchangedNames.bits) {
+      return false;
+    }
+
+    const {
+      bits: { bitmap = '', padding = 0 },
+      hashCount = 0
+    } = unchangedNames;
+
+    if (typeof bitmap === 'string') {
+      const isValidBitmap = this.isValidBase64String(bitmap);
+      if (!isValidBitmap) {
+        logWarn('Invalid base64 string. Applying bloom filter failed.');
+        return false;
+      }
+    }
+
+    const normalizedBitmap = normalizeByteString(bitmap).toUint8Array();
+
+    let bloomFilter;
+    try {
+      // BloomFilter throws error if the inputs are invalid
+      bloomFilter = new BloomFilter(normalizedBitmap, padding, hashCount);
+    } catch (err) {
+      if (err instanceof BloomFilterError) {
+        logWarn('BloomFilter error: ', err);
+      } else {
+        logWarn('Applying bloom filter failed: ', err);
+      }
+      return false;
+    }
+
+    if (bloomFilter.bitCount === 0) {
```

Comment:
What is the reason for adding a short-circuit for the scenario where `bloomFilter.bitCount === 0`? Please add an inline comment about this. Is it because the filterRemovedDocuments() method doesn't work in this case?

Reply:
This was to early return if bitmap is empty.  For example, there are 100 documents, but bitmap is empty, we will run mightContain for 100 times, get all false, and then say applying bloom filter was not successful. 
I can remove this if it is unnecessary.